Add `UNREACHABLE` macro

hebasto opened the pull request:
The `UNREACHABLE` macro was suggested in https://github.com/bitcoin/bitcoin/pull/24812, but during reviewing it has been [dropped](https://github.com/bitcoin/bitcoin/pull/24812#discussion_r851597471):
> This is unused and on a second thought I wonder if there is any value in it. The current code can already use `assert(false)` just fine. Introducing another way to write `assert(false)` will just lead to bike-shedding without any benefit.

This macro prevents compiler warnings -- `-Wreturn-type` (GCC) and `C4715` (MSVC) -- when building for/on Windows.

hebasto commented:
cc @aureleoules @MarcoFalke

MarcoFalke commented on src/util/check.h:
```diff
@@ -95,4 +95,14 @@ T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] const char* f
+ */
+#define UNREACHABLE()                                                                                                                                   \
+    do {                                                                                                                                                \
+        std::cerr << format_internal_error("Unreachable code reached (fatal, aborting)", __FILE__, __LINE__, __func__, PACKAGE_BUGREPORT) << std::endl; \
```
Maybe move the implementation/formatting to the cpp file?

hebasto replied:
Is it worth for a line of code?

hebasto replied:
This approach requires that `libbitcoinconsensus.la` being linked against `libbitcoin_util.a`. The latter, in turn, must be converted into a Libtool archive.

Therefore, I'd prefer to keep this in the header.

aureleoules commented:
Concept ACK

hebasto commented:
> because Windows compilers are broken

They are not broken, as they do not violate the C++ standard. Of course, the way how they consider warnings for `int f(){ assert(false); }` differs from other popular compilers.

> We'll have an unintuitive split where `assert()` is used, _except_ in a switch, because we're trying to suppress randomly (the warnings aren't emitted consistently from switches with mingw-w64) happening warnings.

For non-Windows targets warnings are inconsistent between GCC and Clang as well (remove `assert(false);` statement to see it).

> If we were going to introduce an unreachable macro, we could at least use `__builtin_unreachable` (GCC, Clang, ICC). Of course MSVC doesn't implement that, and instead has `__assume()` (not to be confused with our own `Assume()`).

That could be like an implementation of [`std::unreachable`](https://en.cppreference.com/w/cpp/utility/unreachable) from C++23.

But the suggested macro is safer when it comes to handle, for example, non-defined-explicitly enumeration values, as `std::abort` is better than UB.

hebasto commented:
Updated 05ffaf4591911b3c19ce6ab347538642765190c0 -> 39d681e1c741880f96ef5c2f4c0ae2e93e5ddbcc ([pr26504.01](https://github.com/hebasto/bitcoin/commits/pr26504.01) -> [pr26504.02](https://github.com/hebasto/bitcoin/commits/pr26504.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26504#discussion_r1023719330)

MarcoFalke commented on src/util/check.h:
```diff
@@ -95,4 +95,12 @@ T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] const char* f
         format_internal_error("Unreachable code reached (non-fatal)", \
                               __FILE__, __LINE__, __func__, PACKAGE_BUGREPORT))
 
+[[noreturn]] void inline_unreachable(const char* file, int line, const char* func);
```
nit: I don't think this is inline. Maybe call it `unreachable_fail` for symmetry?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26504#issuecomment-1317292167).

fanquake commented on src/netaddress.cpp:
```diff
@@ -695,7 +696,7 @@ uint32_t CNetAddr::GetLinkedIPv4() const
         return ~ReadBE32(Span{m_addr}.last(ADDR_IPV4_SIZE).data());
     }
-    assert(false);
+    UNREACHABLE();
```
This is now also changing code that does not use switch statements, and I assume doesn't warn under mingw-w64, becoming inconsistent with the rest of the codebase. If we are going to change some of these, we should be changing all, otherwise this is just random refactoring.

hebasto replied:
> I assume doesn't warn under mingw-w64

It does when compiling with `-O0`.

MarcoFalke replied:
There'd be test/lint/lint-assertions.py if this is a goal

hebasto replied:
@fanquake 
> we should be changing all, otherwise this is just random refactoring.

In that case, what macro name could be appropriate?

hebasto commented:
Updated 39d681e1c741880f96ef5c2f4c0ae2e93e5ddbcc -> dd47e2aaf869806455289d55222dd981ee8cea3f ([pr26504.02](https://github.com/hebasto/bitcoin/commits/pr26504.02) -> [pr26504.03](https://github.com/hebasto/bitcoin/commits/pr26504.03)):

- rebased due to conflict with #25112
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26504#discussion_r1024087900)

MarcoFalke commented on src/util/check.cpp:
```diff
@@ -27,3 +28,9 @@ void assertion_fail(const char* file, int line, const char* func, const char* as
+
+void unreachable_fail(const char* file, int line, const char* func)
+{
+    std::cerr << strprintf("Internal bug detected: Unreachable code reached (fatal, aborting)\n%s:%d (%s)\nPlease report this issue here: %s\n", file, line, func, PACKAGE_BUGREPORT) << std::endl;
```
rebase and use the existing string template?

hebasto replied:
[Done](https://github.com/bitcoin/bitcoin/pull/26504#issuecomment-1345356487).

hebasto commented:
Updated dd47e2aaf869806455289d55222dd981ee8cea3f -> 426d6bed7318ad38e0932498d3730af04993ca94 ([pr26504.03](https://github.com/hebasto/bitcoin/commits/pr26504.03) -> [pr26504.04](https://github.com/hebasto/bitcoin/commits/pr26504.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26504#discussion_r1045102086)

To make this PR compatible with the recently merged #26645, the `CopyrightHolders()` and `LicenseInfo()` functions have been moved from `clientversion.{h,cpp}'` to `util/string.{h,cpp}`.

hebasto commented:
Rebased on top of the #26688 and drafted until the latter is being reviewed.

hebasto committed:
util: Add `UNREACHABLE` macro

Co-authored-by: Aurèle Oulès <aurele@oules.com>
Co-authored-by: MacroFake <falke.marco@gmail.com>

hebasto commented:
> @hebasto Let us know if you'll pick this up again or if you want someone else to do it?

Almost ready :)

hebasto commented:
re https://github.com/bitcoin/bitcoin/pull/26504#discussion_r1263587638
> Maybe drop this change and the one to `src/script/interpreter.cpp` for now and only use the second and third commit?

Done. Rebased.

maflcko reviewed (approved):
Nice. Review lgtm

jonatack reviewed (commented):
If this is done, IIUC the developer notes should be updated.

hebasto committed:
Use `UNREACHABLE` macro in non-RPC code

This change prevents `-Wreturn-type` and `C4715` warnings when building
for/on Windows.

hebasto commented on src/net.h:
```diff
@@ -493,7 +493,7 @@ class CNode
                 return true;
         } // no default case, so the compiler can warn about missing cases
 
```
Done.

hebasto commented:
@MarcoFalke 
> Also, could update title, description and commit message?

The title an the description have been updated. What update for commit messages would you like to see?

hebasto committed:
doc: Amend enumerations chapter in Developer Notes

hebasto commented on doc/developer-notes.md:
````diff
@@ -844,11 +844,11 @@ int GetInt(Tabs tab)
 ```
 
-*Rationale*: The comment documents skipping `default:` label, and it complies with `clang-format` rules. The assertion prevents firing of `-Wreturn-type` warning on some compilers.
+*Rationale*: The comment documents skipping `default:` label, and it complies with `clang-format` rules. The `UNREACHABLE` macro prevents firing of `-Wreturn-type`/`C4715` warning on some compilers. In RPC code the `NONFATAL_UNREACHABLE` macro might be more appropriate instead.
````
Thanks! Updated.

fanquake commented:
> Apparently, this macro prevents compiler warnings

Either it does or it doesn't? If it doesn't, then any workarounds should be removed, and warnings re-enabled.

If it doesn't, there's no point to this refactoring at all, if it doesn't even acheive the stated goals.

> I find it frustrating to play compiler golf against g++ in the CI on every code push that uses a switch-case on an enum class.

Which CI is this? The only one where it is an issue has the warnings disabled?

fanquake commented:
> but I think having a macro for those that find it helpful, can't hurt, can it?

I agree that having a macro could be useful.

Separately, note that upstream, the missing `noreturn` attributes have been added to `assert` in the mingw-w64 headers, https://github.com/mingw-w64/mingw-w64/commit/1690994f515910a31b9fb7c7bd3a52d4ba987abe, so beginning with [11.0.0](https://www.mingw-w64.org/changelog/), there are no-longer `-Wreturn-type` warnings with `assert(false)` usage. See #28092 where I've updated the docs.

fanquake commented:
> I don't care about updating the existing code,

To be clear, I agree, and am still a NACK to any code changes here.

hebasto commented:
> Separately, note that upstream, the missing `noreturn` attributes have been added to `assert` in the mingw-w64 headers, [mingw-w64/mingw-w64@1690994](https://github.com/mingw-w64/mingw-w64/commit/1690994f515910a31b9fb7c7bd3a52d4ba987abe), so beginning with [11.0.0](https://www.mingw-w64.org/changelog/), there are no-longer `-Wreturn-type` warnings with `assert(false)` usage.

That change has been reverted here: https://github.com/mingw-w64/mingw-w64/commit/ecf2328a328d11dec7044b40b2b5e93b5b2b9d9e.
